File: Gombang/[LeetCode75] - [DFS]/104_Maximum_Depth_of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    int _maxDepth = 0;

    int maxDepth(TreeNode* root) {
        DFS(root, 1);

        return _maxDepth;
    }

    void DFS(TreeNode* node, int depth)
    {
        if (node == nullptr)
            return;

        _maxDepth = max(_maxDepth, depth);
        DFS(node->left, depth + 1);
        DFS(node->right, depth + 1);
    }
};

File: Seori/LeetCode 75/Binary Tree_DFS/104. Maximum Depth of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        
        def dfs(root: Optional[TreeNode], depth: int) -> int:
            # 재귀 종료 조건 : 리프에 도착했을 때
            if root == None: 
                return depth - 1

            # 트리 전위순회
            left = dfs(root.left, depth + 1)
            right = dfs(root.right, depth + 1)
            return max(left, right)

        return dfs(root, 1)
        